chain/consensus/tendermint/app.go

```diff
@@ -0,0 +1,171 @@
+package tendermint
+
+// This is an adaptation of the reference Tendermint application for Eudico needs.
+// The reference implementation can be found here - https://github.com/tendermint/tendermint/tree/master/abci/example.
+
+import (
+	"time"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	logger "github.com/tendermint/tendermint/libs/log"
+)
+
+const (
+	codeBadRequest   = 513 // arbitrary non-zero
+	codeStateFailure = 514
+)
+
+var (
+	version                  = "0.0.3"
+	_       abci.Application = (*Application)(nil)
+)
+
+type Application struct {
+	consensus *State
+	logger    logger.Logger
+}
+
+func NewApplication() (*Application, error) {
+	return &Application{
+		consensus: NewState(),
+		logger:    logger.NewNopLogger(),
+	}, nil
+}
+
+func (a *Application) Info(abci.RequestInfo) (resp abci.ResponseInfo) {
+	return abci.ResponseInfo{
+		Data:             time.Now().Format("2006.01.02 15:04:05"),
+		Version:          version,
+		LastBlockHeight:  a.consensus.Height(),
+		LastBlockAppHash: a.consensus.Hash(),
+	}
+}
+
+func (a *Application) InitChain(req abci.RequestInitChain) (resp abci.ResponseInitChain) {
+	return abci.ResponseInitChain{}
+}
+
+func (a *Application) Query(req abci.RequestQuery) (resp abci.ResponseQuery) {
+	switch req.Path {
+	case "/reg":
+		subnetID := req.Data
+		height, ok := a.consensus.GetSubnetOffset(subnetID)
+		if !ok {
+			return abci.ResponseQuery{
+				Code: codeStateFailure,
+				Log:  "subnet offset hasn't been set yet",
+			}
+		}
+
+		regResp := RegistrationMessageResponse{
+			Name:   subnetID,
+			Offset: height,
+		}
+
+		data, err := regResp.Serialize()
+		if err != nil {
+			return abci.ResponseQuery{
+				Code: codeBadRequest,
+				Log:  err.Error(),
+			}
+		}
+
+		resp.Key = req.Data
+		resp.Value = data
+		return
+	default:
+		return abci.ResponseQuery{
+			Code: abci.CodeTypeOK,
+		}
+	}
+}
+
+func (a *Application) BeginBlock(req abci.RequestBeginBlock) (resp abci.ResponseBeginBlock) {
+	return abci.ResponseBeginBlock{}
+}
+
+func (a *Application) CheckTx(req abci.RequestCheckTx) (resp abci.ResponseCheckTx) {
+	tx := req.GetTx()
+
+	_, code, err := parseTx(tx)
+	if err != nil {
+		return abci.ResponseCheckTx{
+			Code: code,
+			Log:  err.Error(),
+		}
+	}
+
+	return abci.ResponseCheckTx{
+		Code: abci.CodeTypeOK,
+	}
+}
+
+func (a *Application) DeliverTx(req abci.RequestDeliverTx) (resp abci.ResponseDeliverTx) {
+	tx := req.GetTx()
+
+	msg, code, err := parseTx(tx)
+	if err != nil {
+		return abci.ResponseDeliverTx{
+			Code: code,
+			Log:  err.Error(),
+		}
+	}
+
+	switch subnet := msg.(type) {
+	case *RegistrationMessageRequest:
+		height := a.consensus.SetOrGetSubnetOffset(subnet.Name)
+		regResp := RegistrationMessageResponse{
+			Name:   subnet.Name,
+			Tag:    subnet.Tag,
+			Offset: height,
+		}
```

Comment:
Yes, you are right. We register a Tendermint blockchain offset for a new subnet here.
Suppose we just started a root subnet and Tendermint blockchain, then its offset will be for example 3 blocks.
And then after an hour, we decided to register a new subnet, its offset will be 1000. If we don't use offsets then the second subnet will have to access all previous Tendermint blocks. 

> Why I ask is because I am wondering how critical would it be for a rogue node to register a subnet he is not part of (or that doesn't even exist yet).
It is critical. A rogue node can register a network earlier than it would be registered, e.g. at epoch h. Then when the network is instantiated by real participants at epoch H they will have to process all blocks from epoch h to H.

Reply:
Fixed. Now only one subnet can be registered in Tendermint. It doesn't protect against DoS attacks but mitigates them.